jwk/jwk.go

```diff
@@ -169,59 +194,59 @@ func PublicKeyOf(v interface{}) (interface{}, error) {
 	}
 }
 
-// Fetch fetches a JWK resource specified by a URL
-func Fetch(urlstring string, options ...Option) (*Set, error) {
-	u, err := url.Parse(urlstring)
+// Fetch fetches a JWK resource specified by a URL. The url must be
+// pointing to a resource that is supported by `net/http`.
+//
+// If you would like to refresh the jwk.Set from a remote resource,
+// consider using jwk.AutoRefresh, which automatically refreshes
+// jwk.Set objects asynchronously.
+func Fetch(ctx context.Context, urlstring string, options ...FetchOption) (Set, error) {
+	res, err := fetch(ctx, urlstring, options...)
 	if err != nil {
-		return nil, errors.Wrap(err, `failed to parse url`)
+		return nil, err
 	}
 
-	switch u.Scheme {
-	case "http", "https":
-		return FetchHTTP(urlstring, options...)
-	case "file":
-		f, err := os.Open(u.Path)
-		if err != nil {
-			return nil, errors.Wrap(err, `failed to open jwk file`)
-		}
-		defer f.Close()
-
-		return Parse(f)
+	defer res.Body.Close()
+	keyset, err := ParseReader(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, `failed to parse JWK set`)
 	}
-	return nil, errors.Errorf(`invalid url scheme %s`, u.Scheme)
-}
-
-// FetchHTTP wraps FetchHTTPWithContext using the background context.
-func FetchHTTP(jwkurl string, options ...Option) (*Set, error) {
-	return FetchHTTPWithContext(context.Background(), jwkurl, options...)
+	return keyset, nil
 }
 
-// FetchHTTPWithContext fetches the remote JWK and parses its contents
-func FetchHTTPWithContext(ctx context.Context, jwkurl string, options ...Option) (*Set, error) {
+func fetch(ctx context.Context, urlstring string, options ...FetchOption) (*http.Response, error) {
 	httpcl := http.DefaultClient
+	bo := backoff.Null()
 	for _, option := range options {
 		switch option.Ident() {
 		case identHTTPClient{}:
 			httpcl = option.Value().(*http.Client)
+		case identFetchBackoff{}:
+			bo = option.Value().(backoff.Policy)
 		}
 	}
 
-	req, err := http.NewRequest(http.MethodGet, jwkurl, nil)
+	req, err := http.NewRequest(http.MethodGet, urlstring, nil)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to new request to remote JWK")
 	}
 
-	res, err := httpcl.Do(req.WithContext(ctx))
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to fetch remote JWK")
-	}
-	defer res.Body.Close()
+	b := bo.Start(ctx)
+	var lastError error
+	for backoff.Continue(b) {
+		res, err := httpcl.Do(req.WithContext(ctx))
```

Comment:
> I think that with crypto/ed25519 you need at least 1.13

If you only want to support Go 1.13 or later, you can change it.

```diff
-       req, err := http.NewRequest(http.MethodGet, urlstring, nil)
+       req, err := http.NewRequestWithContext(ctx, http.MethodGet, urlstring, nil)
        if err != nil {
                return nil, errors.Wrap(err, "failed to new request to remote JWK")
        }
@@ -234,7 +234,7 @@ func fetch(ctx context.Context, urlstring string, options ...FetchOption) (*http
        b := bo.Start(ctx)
        var lastError error
        for backoff.Continue(b) {
-               res, err := httpcl.Do(req.WithContext(ctx))
+               res, err := httpcl.Do(req)
                if err != nil {
                        lastError = errors.Wrap(err, "failed to fetch remote JWK")
                        continue
```

Reply:
Fixed. I was using `http.NewRequestWithContext()` in tests anyway